prov/gnix: Reference counting with MR system

jswaro opened the pull request:
NICs allocated from the context of a memory registration start with
a single reference that is never dropped because a second reference
is taken by the original registration. gnix_nic_alloc assumes the
calling function will take the reference provided by the function
and that assumption wasn't being observed.

This mod fixes the issue by only taking a reference if a NIC wasn't
created during the memory registration request.

Signed-off-by: James Swaro <jswaro@cray.com>

@sungeunchoi @ztiffany 

closes #1086

sungeunchoi commented:
How did you test this?  Can we add something to help check for this in our unit tests?

ztiffany commented:
What if we moved the _gnix_nic_alloc to the point where we actually do the registration call?  Looks like that's in __gnix_generic_register.

jswaro commented:
> How did you test this? Can we add something to help check for this in our unit tests?

I tested this with gnitest using the 'mr_internal_bare/basic_init' test. I scanned the debug log for the destructor call with and without the patch. I also added code to the gnix_mr_reg function at the bottom of the function call to print the reference count and verify that it had dropped the original nic reference when the registration had been closed. 

We could probably add something, but it would require maintenance since we normally don't inspect provider structures.

> What if we moved the _gnix_nic_alloc to the point where we actually do the registration call? Looks like that's in __gnix_generic_register. 

What is the benefit of moving it?

sungeunchoi commented:
> We could probably add something, but it would require maintenance since we normally don't inspect provider structures.

That's not really true.  We look poke inside the provider structures all over the place.  I think since you found this problem, it's worth putting in a test.

jswaro commented:
I'll add a test.

jswaro commented:
I'll squash prior to merge.

ztiffany commented:
The way the code is structured, it doesn't seem like the NIC you create is necessarily the NIC used for the registration.  I'm also seeing refs taken on NIC and domain in two places for each registration.  Is that intentional?

jswaro commented:
> The way the code is structured, it doesn't seem like the NIC you create is necessarily the NIC used for the registration.

The NIC is only created if there are no NICs available. This would be the only NIC the registration could be tied to. 

> I'm also seeing refs taken on NIC and domain in two places for each registration. Is that intentional?

The reference to the NIC and domain was taken twice originally because the memory registration object passed to the user and the memory registration object owned by the provider were two distinct objects. The user object could be closed, but the internally cached registration would still exist. The upside is that we only take two references if there isn't an existing registration which would cover the requested memory region. The downside is that this is still required to support lazy deregistration, because the internal provider object may exist long after the user object has been released.

ztiffany commented:
>The NIC is only created if there are no NICs available. This would be the only NIC the registration could be tied to.

If an EP was being created by another thread,  you could have problems.  There are bad assumptions being made here.

ztiffany commented:
> The reference to the NIC and domain was taken twice originally because the memory registration object passed to the user and the memory registration object owned by the provider were two distinct objects. 

If that's changed, why not update the ref counting?

jswaro commented:
> If that's changed, why not update the ref counting?

I'll reiterate. 

The downside is that this is still required to support lazy deregistration, because the internal provider object may exist long after the user object has been released. 

> If an EP was being created by another thread, you could have problems. There are bad assumptions being made here.

That's a threading issue. See #1098.

jswaro commented:
FWIW, I understand your concern about creating one NIC and finding a different one by the time registration has occurred. 

Given that the list is traversed in the order that the NICs have been inserted, I doubt that ever happens, but it's worth fixing just to be sure that we don't have a reference on two separate NICs without having having a pointer to both NICs. 

I'll fix it in a different PR. This PR is fixing the first point you raised in the issue.
>gnix_nic_alloc in gnix_mr.c adds a references on the newly created NIC that would never be removed.

The second point is being addressed in #1098.

I'll address this last issue in a separate issue, #1102.

ztiffany commented:
>Given that the list is traversed in the order that the NICs have been inserted

We have code now that reorders the NIC list.

jswaro commented:
> We have code now that reorders the NIC list.

Thanks for the heads up. I'll address that in the PR for 1102.

sungeunchoi commented:
Thx for test code.  👍

jswaro commented:
Closing this to merge with 1103
